### DISCUSSION TITLE
[RFC] Warn when an installed package is not compatible with the managed workflow

### DISCUSSION THREAD

**Discussion by brentvatne** [expo/expo#8167] [Proposals and Ideas]
Maintainers of React Native libraries like react-native-localize and others with native dependencies often have to field questions about why their library is not working in a managed Expo project ([example](https://github.com/react-native-community/react-native-localize/issues/30#issuecomment-510520508)). Some related thoughts:

- **It takes time for people to form a mental model for how React Native works**. A lot of developers come from exclusively a web development background and they don't immediately grok the concept of native modules or that a npm package includes code that isn't JS code.
- **It's not easy to determine if a library has native code that the user needs to link before using it**. Example: [this](https://github.com/facebook/react-native-website/pull/1902#discussion_r420303539) is the best we can do currently in the "Using Libraries" docs in React Native. We could perhaps use some kind of heuristic, eg: via @zoontek "I would go with presence a check of `*.{java,m,mm,cpp,swift}` files, excluding possible examples directory (never at root, contains a package.json with react-native as dependency)." What false positives would this lead to? What about false negatives? This would be useful to have in React Native upstream ("hey, you installed a lib with native dependencies for iOS and Android, be sure to run `pod install` and rebuild your project before using it!"), as well as in Expo CLI.
- **If it does have native code, does it work with Expo managed workflow?** - some libraries, eg: react-native-webview, do indeed. They are listed in [the API reference](https://docs.expo.io/versions/latest/). Folks can also refer to [React Native Directory](https://reactnative.directory/) and filter by supported in Expo client.
- **We don't clearly document an algorithm developers can use to determine if a library is compatible with managed workflow** - so they learn by trial and error. I am working on a page for our documentation right now to help with this.
- **When do we warn users?** most libraries suggest using `npm install` or `yarn add`, so if they are following their READMEs then they will do that and we don't have a chance to hook into it via `expo install` and warn them. When do we check the project and warn? Adding a `postinstall` script feels too invasive and easy for people to just delete without understanding why it's there.
- **Is there anything we can do at runtime?** when a native module is missing the current error message is not very clear to users - something like "undefined is not an object" or "Native module is null" or "View manager cannot be null". If we could detect errors caused by accessing native modules / native module properties that don't exist and describe the issue more clearly / link people to more information, this could be a great fallback.

## The goals of all of this are the following

- Developers are guided through building a mental model of how React Native works and what they need to consider when installing packages for the managed workflow.
- When they misunderstand, they are met with useful warnings along the way and ultimately a useful error screen if they don't heed the warnings.
- Maintainers of libraries like react-native-localize will have less issues to deal with around installation problems in managed Expo projects and, when they do encounter them, they can link them to an easy to follow resource to understand what is wrong and then just close the issue.

**Comment by brentvatne** [comment]
I've added a first pass at a guide for using libraries in managed and bare workflows: https://github.com/expo/expo/commit/a9ac975ce0d264a8599b5f20538ee1c4bbea2160

This includes suggestions for how developers can determine if a library is compatible with the Expo managed workflow or not. One thing that React Native Community folks could help out with is to [populate React Native Directory with their libraries](https://github.com/react-native-community/directory#how-do-i-add-a-library) to help users quickly determine whether or not it is compatible with Expo managed workflow as per the suggestions in the docs.

**Reply by brentvatne** [reply]
Deployed URL is https://docs.expo.io/workflow/using-libraries/

**Comment by zoontek** [comment]
It's also possible to improve the React Native documentation page to reduce the ambiguity between bare RN and Expo. Currently, both are mixed up in the [Getting started section](https://reactnative.dev/docs/environment-setup).

I suggest a clear choice at start, then a complete tab with information about the selected option, and only this one. It also could be nice to replace the occurrences of "Your React Native app" with "Your Expo app" in here.

![Screenshot 2020-05-06 at 19 42 26](https://user-images.githubusercontent.com/1902323/81210143-c38dac00-8fd1-11ea-955a-3ff78b6ddd3a.png)

_(It's just for the idea, the design is not the best / the more clear possible)_

Under the **Now what?** title, it could be nice to add an illustration with a link to redirect newcomers to https://docs.expo.io/tutorial/planning/. The current Expo docs link is not visible enough, and people might miss the Expo docs entirely and all what it offers (the internal modules).

**Reply by brentvatne** [reply]
![Screen Shot 2020-05-06 at 12 35 40 PM](https://user-images.githubusercontent.com/90494/81220408-1b0e2680-8f96-11ea-97fb-9ac46fe86503.png)

above is the current state of the "Setting up the development environment" page. this is the result of some discussion with the react native website maintainers and we had a similar goal - to reduce the ambiguity. so we wanted to lead with some information about each. i wonder if we could achieve a similar thing to what you have mocked up above by changing the color of the bolded text so the attention is drawn more towards reading that and less to the large black codeblocks

![Screen Shot 2020-05-06 at 12 38 01 PM](https://user-images.githubusercontent.com/90494/81220645-822bdb00-8f96-11ea-9b75-a8c99f0d55d0.png)

good point about the "Now what?" section! I'll take a stab at clarifying it further when I have a few minutes free.

we also include caveats about custom native modules below the "Now what?" but 1) it's not clear to people at this point what a custom native module is, or even really what a native module is 2) we can't depend on everyone reading docs pages in full.

**Reply by zoontek** [reply]
I know for sure that a lot of people skip huge parts of documentations 😅. Maybe the "native module error" in Expo could give a short link to a page with caveats, basic explanations & possible solutions (check for similar API in Expo modules, ejecting, switch to bare RN)?

**Reply by brentvatne** [reply]
yeah i think we'll need to have that as a mitigation in the worst case! i created this issue and assigned myself https://github.com/expo/expo/issues/8163. we won't be able to ship it until sdk38 (end of june) but that gives me time to figure out a good solution and implement it anyhow
